sdks/java/core/src/main/java/org/apache/beam/sdk/util/PathUtils.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.util;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.Strings;
+import java.net.URI;
+import java.net.URISyntaxException;
+import javax.annotation.Nonnull;
+
+/**
+ * Utility class for handling file paths.
+ */
+public class PathUtils {
+
+  private static final String URI_DELIMITER = "/";
+
+  /**
+   * Resolve multiple {@code others} against the given {@code directory} sequentially.
+   *
+   * <p>Sees {@link #resolveAgainstDirectory} for the differences with {@link URI#resolve}.
+   *
+   * @throws IllegalArgumentException if others contains {@link URI} query or fragment components.
+   */
+  public static String resolveAgainstDirectory(
+      @Nonnull String directory, @Nonnull String other, @Nonnull String... others) {
```

Comment:
Annotating as `Nonnull` is unnecessary - everything is non-null unless annotated as `Nullable`. This is the style followed by the rest of Beam codebase.